### PR TITLE
[IMP] payment: filter out acquirers based on the amount

### PR DESCRIPTION
**Pull request by hote-odoo**
This commit adds the possibility to define a maximum payment amount that
a given acquirer can process. If the payment amount exceeds the value,
the acquirer is filtered out of the available acquirers listed on the
payment forms.

While we're at it, the field `country_ids` is renamed to
`available_country_ids` to better depict that it is not a property of
the acquirer, but a configuration option. It will also be coherent with
the field `available_currency_ids` that is expected to be added soon.

Task - 2162165

See also:
- https://github.com/odoo/enterprise/pull/24412
- https://github.com/odoo/documentation/pull/1835
- https://github.com/odoo/upgrade/pull/3703

### CODE REVIEW

**Comment by hote-odoo** [comment]
Still working on it.

File: addons/payment/models/payment_acquirer.py
Comment (fmdl) [inline addons/payment/models/payment_acquirer.py:145]: it is ugly, because the result of the partner search can return million of partner.
I have add sudo to allow read restricted field.

```python
def is_allowed_partner(self, partner_id):
      self.ensure_one()
      return bool(self.env['res.partner'].sudo().browse(partner_id).filtered_domain(safe_eval(self.payment_domain or '[]')))
```

Comment (hote-odoo) [reply addons/payment/models/payment_acquirer.py:145]: thank you for the suggestion, I will implement some of it

File: addons/payment/models/payment_acquirer.py
Comment (hote-odoo) [reply addons/payment/models/payment_acquirer.py]: I went to see the source of `safe_eval` and it better to use it indeed

Comment (odony) [reply addons/payment/models/payment_acquirer.py]: @fmdl `ast.literal_eval()` is more restrictive than a full `safe_eval()`, it's roughly similar to `safe_eval.const_eval()`, so it's ok to use it for e.g. static domains built with ... literals :)

I'll leave it to the team to review the whole PR when it's ready (still draft ATM)

Comment (hote-odoo) [reply addons/payment/models/payment_acquirer.py]: guess I will go back to `ast.literal_eval()` then

**Review by chevalierv — CHANGES_REQUESTED** [review]
Thanks for your work on this task!

After further analysis, I don't think that you're following the best path. I've left a comment to guide you to a better solution (IMO), that should let you drop your changes in all views.

File: addons/payment/models/payment_acquirer.py
Comment (hote-odoo) [reply addons/payment/models/payment_acquirer.py]: done :)

File: addons/payment/tests/test_acquirer.py
Comment (chevalierv) [inline addons/payment/tests/test_acquirer.py:21]: Maybe you could add a test that no acquirer is returned when the amount is zero?

File: addons/payment/models/payment_acquirer.py
Comment (chevalierv) [inline addons/payment/models/payment_acquirer.py:371]: I think it will be clearer that way. 
```suggestion
            domain = expression.AND([
                 domain, [ '|', 
                    ('maximal_amount', '>=', converted_amount), 
                    '|', ('maximal_amount', '=', False), ('maximal_amount', '=', 0.00)
                 ]
            ])
```

**Review by chevalierv — APPROVED** [review]

File: addons/payment/models/payment_acquirer.py
Comment (hote-odoo) [reply addons/payment/models/payment_acquirer.py]: done :)

File: addons/payment/tests/test_acquirer.py
Comment (hote-odoo) [reply addons/payment/tests/test_acquirer.py:21]: done :)

**Comment by AntoineVDV** [comment]
Dismissing reviews from rd-sales and rd-ecommerce since they were assigned for a single line of diff in the payment portal.

**Review by AntoineVDV — APPROVED** [review]
@robodoo r+

**Pull request closed without merging**
